### ISSUE TITLE
Support tsconfig references

### ISSUE THREAD

**Issue by molvqingtai** [typescript-eslint/typescript-eslint#4094]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
**.eslintrc.cjs**

```js
{
  env: {
    browser: true,
    es2021: true,
    node: true
  },
  extends: ['standard', 'eslint-config-standard-with-typescript', 'prettier'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaVersion: 12,
    sourceType: 'module',
    project: './tsconfig.json'
  },
  plugins: ['@typescript-eslint', 'prettier'],
  rules: {
    'prettier/prettier': 'error'
  }
}
```
**tsconfig.eslint.json**
```
{
  "compilerOptions": {
    "composite": true
  },
  "extends": "./tsconfig.json",
  "include": ["__tests__/**/*.ts"]
}

```

**tsconfig.json**
```JSON
{
  "compilerOptions": {
    "baseUrl": ".",
    "rootDir": "src",
  },
  "include": ["src"],
  "references": [
    {
      "path": "./tsconfig.jest.json"
    },
    {
      "path": "./tsconfig.eslint.json"
    }
  ]
}

```

**Expected Result**
It should work in the same `__tests__` folder as the vscode checker.
https://www.typescriptlang.org/docs/handbook/project-references.html

**Actual Result**
```shell
  0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: __tests__/TypeError.spec.ts.
```

**Additional Info**
It is recommended that "references" be supported, keeping the same behaviour as vscode

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `5.3.0` |
| `TypeScript`                | `4.4.4` |
| `ESLint`                    | `7.32.0` |
| `node`                      | `16.13.0` |

**Comment by jensmeindertsma** [comment]
Yes, I'm experiencing this as well. This used to work on v4, now doesn't anymore.

**Comment by jensmeindertsma** [comment]
This is what the README says:
>  If you use project references, TypeScript will not automatically use project references to resolve files. This means that you will have to add each referenced tsconfig to the `project` field either separately, or via a glob.

This used to work though?

**Comment by JoshuaKGoldberg** [comment]
@molvqingtai @jensmeindertsma could both of you please post a full reproduction repository? We can try to piece together something from the bits of config given but it'd be a lot more reliable to know that it's broken on your end.

**Comment by jensmeindertsma** [comment]
@JoshuaKGoldberg absolutely, here you go: https://github.com/jensmeindertsma/typescript-eslint-project-references

**Comment by jensmeindertsma** [comment]
Let me know if you need anything else @JoshuaKGoldberg 😊

**Comment by bradzacher** [comment]
We have never supported project references - https://github.com/typescript-eslint/typescript-eslint/issues/2094

Project references are not automatic in the TS APIs, and it's a minefield of problems to support them for our usecase.

There are some known bugs with some setups as well - https://github.com/typescript-eslint/typescript-eslint/issues/3851

v5 didn't change anything with our parser packages, so there's nothing that would have broken between the last v4.33 and v5.3.
The last change to the parser packages was in v4.29 (some 3 months ago)

> This is what the README says

That section of the readme has been there for quite a long time (>2yr).

**Comment by bradzacher** [comment]
as mentioned above - we do not and have never support project refs.
Follow #2094 for any future updates on this situation.
